File: src/test/matrix/py/sampling.py
#!/usr/bin/env python3
"""MCP sampling matrix cell server (feature=sampling, language=py, transport=stdio).

Feature under test: SERVER-INITIATED SAMPLING.
A tool handler turns around and asks the CLIENT to sample an LLM via the
sampling/createMessage server->client request (ctx.session.create_message).

Exposes:
  - 1 sampling tool:  ask_llm   (handler calls ctx.session.create_message(...))
  - 1 plain tool:     echo      (no sampling; keeps a trivial tool available)

Notes on expected behavior against mcpgaze:
  mcpgaze acts as a MINIMAL client during snapshot/conform/health: it advertises
  no sampling capability and does NOT answer sampling/createMessage requests.
  Therefore initialize + tools/list MUST still succeed (snapshot lists 2 tools,
  conform --all PASSES, health --once is UP). Calling `ask_llm` will block on a
  client response that never comes, so that tools/call is expected to err/time
  out at the server side -- the key assertion is mcpgaze must NOT crash.

Built on the real Python MCP SDK (FastMCP).
"""

from mcp.server.fastmcp import Context, FastMCP
from mcp.types import SamplingMessage, TextContent

mcp = FastMCP("mcpgaze-sampling-py")


# --- Sampling tool: handler asks the CLIENT to sample an LLM ---------------
@mcp.tool(
    name="ask_llm",
    title="Ask the client's LLM",
    description=(
        "Ask the connected MCP client to sample its LLM with a prompt and "
        "return the model's text. This triggers a server-initiated "
        "sampling/createMessage request."
    ),
)
async def ask_llm(prompt: str, ctx: Context) -> str:
    """Server-initiated sampling: round-trip a prompt through the client's LLM."""
    result = await ctx.session.create_message(
        messages=[
            SamplingMessage(
                role="user",
                content=TextContent(type="text", text=prompt),
            )
        ],
        max_tokens=256,
        system_prompt="You are a helpful assistant invoked via MCP sampling.",
        temperature=0.7,
    )
    content = result.content
    if isinstance(content, TextContent):
        return content.text
    return str(content)


# --- Plain tool (no sampling) ---------------------------------------------
@mcp.tool()
def echo(message: str) -> str:
    """Echo back the provided message (no sampling involved)."""
    return f"echo: {message}"


if __name__ == "__main__":
    mcp.run(transport="stdio")

File: src/test/matrix/ts/sampling.mjs
#!/usr/bin/env node
// sampling-ts matrix cell: MCP server (stdio) whose tool handler issues a
// server->client sampling/createMessage request. Uses the real
// @modelcontextprotocol/sdk. Initialize + tools/list must always succeed; the
// sampling tool only resolves if the CLIENT answers the sampling request.
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";

const server = new McpServer(
  { name: "sampling-ts", version: "1.0.0" },
  { capabilities: { tools: {} } }
);

// The sampling tool: its handler asks the CLIENT to sample an LLM via
// server.server.createMessage(...). If the client (e.g. mcpgaze in minimal-
// client mode) never answers, the request stays pending; we bound it with the
// request's own timeout so the tools/call resolves to a clean error instead of
// hanging forever.
server.registerTool(
  "summarize_via_client",
  {
    title: "Summarize via client sampling",
    description:
      "Ask the connected client to sample an LLM (sampling/createMessage) " +
      "to summarize the provided text, then return the model's reply.",
    inputSchema: {
      text: z.string().describe("Text for the client's LLM to summarize"),
    },
  },
  async ({ text }) => {
    const result = await server.server.createMessage(
      {
        messages: [
          {
            role: "user",
            content: {
              type: "text",
              text: `Summarize the following in one sentence:\n\n${text}`,
            },
          },
        ],
        maxTokens: 200,
      },
      // Bound the wait so a non-answering client yields an error, not a hang.
      { timeout: 8000 }
    );

    const reply =
      result?.content?.type === "text"
        ? result.content.text
        : JSON.stringify(result?.content ?? result);

    return { content: [{ type: "text", text: `model said: ${reply}` }] };
  }
);

// A plain tool so tools/list works and is meaningful even without sampling.
server.registerTool(
  "echo",
  {
    title: "Echo",
    description: "Echo back the provided message (no sampling involved)",
    inputSchema: { message: z.string().describe("Message to echo back") },
  },
  async ({ message }) => ({
    content: [{ type: "text", text: `echo: ${message}` }],
  })
);

const transport = new StdioServerTransport();
await server.connect(transport);
